HANA database driver

mrylov opened the pull request:
Dear GDAL/OGR community,

We, the SAP HANA Spatial Team, would like to add HANA support to the GDAL library. [SAP HANA](https://www.sap.com/products/hana.html) is an in-memory database with an [OGC-compliant](http://www.opengeospatial.org/resource/products/details/?pid=1303) [spatial engine](https://help.sap.com/viewer/cbbbfc20871e4559abfd45a78ad58c02/2.0.04/en-US/e1c934157bd14021a3b43b5822b2cbe9.html). A free community edition of SAP HANA is available [here](https://www.sap.com/cmp/td/sap-hana-express-edition.html).

Our [previous attempt](https://lists.osgeo.org/pipermail/gdal-dev/2021-July/054460.html) to get an approval from the PSC was not successful. For some reason, the PSC Members were rather inactive, except Even, and the decision has not been made. Therefore, we decided to take a chance with a pull request, hoping to find a community reviewer who will be interested to do the review.

### Content 
This pull request includes:
 - Implementation of the driver
 - Driver documentation
 - Python tests
 - CI integration using a HANA Cloud instance

### Requirements  
The HANA driver requires the open-source [odbc-cpp-wrapper](https://github.com/SAP/odbc-cpp-wrapper) library that is going to be used only during the compilation phase and linked statically. Thus, end users will get only one dynamic/shared library of the HANA driver.

### License
The HANA driver is licensed under an [MIT/X style license](https://github.com/OSGeo/gdal/blob/master/LICENSE.TXT#L12).

### Responsibilities
We promise to keep tracking of new issues/bugs, CI runs, API changes, documentation 
improvements related to the HANA driver and make sure that they are fixed/implemented promptly.
There will be a contact person at SAP responsible for the maintenance.

rouault commented:
The project is discussing a RFC about policies for substantial code additions: https://github.com/OSGeo/gdal/pull/5128

mkemeter commented:
Thanks for the notification! We will add our perspective to the discussion.

mrylov commented:
We didn't expect to see both Github Actions and Travis CI enabled in this repo. We are working on resolving the issues on Travis-CI. Sorry for the inconvenience caused by that.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanalayer.cpp:
```diff
@@ -0,0 +1,955 @@
+
+GIntBig OGRHanaLayer::GetFeatureCount(CPL_UNUSED int force)
+{
+    GIntBig ret = FALSE;
```
```suggestion
    GIntBig ret = 0;
```

mrylov replied:
Fixed in 644e57d53b752d4ef164e70e9424b3bc73246580.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanalayer.cpp:
```diff
@@ -0,0 +1,955 @@
+            if (srs_ != nullptr)
+                srs_->Reference();
+            else
+                srid_ = -1;
```
useless

rouault commented on ogr/ogrsf_frmts/hana/ogrhanalayer.cpp:
```diff
@@ -0,0 +1,955 @@
+/*                           GetSpatialRef()                            */
+/************************************************************************/
+
+OGRSpatialReference* OGRHanaLayer::GetSpatialRef()
```
implementing that method isn't needed if you SetSpatialRef() on created OGRGeometryFieldDefn.  And actually you should make sure that a user calling GetGeomFieldDefn(0)->GetSpatialRef() gets the same result as below. (didn't check if it is already the case. Lost in review...)

mrylov replied:
Thanks for pointing this out. This function was removed in 0d769d4a7dcaab9327c6bcd69042ddb04decddba.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+OGRHanaTableLayer::OGRHanaTableLayer(OGRHanaDataSource* datasource, int update)
+    : OGRHanaLayer(datasource)
+    , updateMode_(update)
+    , batchSize_(4 * 1024)
```
I believe the doc says the default batch size is 1 KB not 4 KB

mrylov replied:
Good catch. The value was changed to 4MB in d14c72057258326c1e697eb32715f605ba622237.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+                    statement.setInt(paramIndex, odbc::Int());
+                else
+                {
+                    if (!CanCastIntBigTo<std ::int32_t>(feature->GetFID()))
```
```suggestion
                    if (!CanCastIntBigTo<std::int32_t>(feature->GetFID()))
```

mrylov replied:
Reworked in c9e8b4bf2dcc5962db78e8063d5c01c8e152b931.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+                    statement.setLong(paramIndex, odbc::Long());
+                else
+                {
+                    if (!CanCastIntBigTo<std::int64_t>(feature->GetFID()))
```
this shouldn't happen . GIntBig == std::int64_t

mrylov replied:
Removed in c9e8b4bf2dcc5962db78e8063d5c01c8e152b931.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+    if (!newFeature)
+    {
+        ++paramIndex;
+        if (!CanCastIntBigTo<std::int64_t>(feature->GetFID()))
```
shouldn't happen

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+        OGRErr err = GetGeometryWkb(feature, static_cast<int>(i), wkb);
+        if (OGRERR_NONE != err)
+            return err;
+        statement.setBytes(paramIndex, wkb.data, wkb.size);
```
is wkb.data CPLFree()'d after usage ?

mrylov replied:
No, there is no need in doing this.
HANA layers has an internal member  std::vector<char> dataBuffer_, that is used for reading/writing of binary/string data. The wkb.data member just points to dataBuffer_.data().

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+        return OGRERR_FAILURE;
+    }
+
+    if (nullptr == feature)
```
already tested above

mrylov replied:
Removed in f2f6378e7df638dcb28936709e56af9c1c25127d.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+        return OGRERR_FAILURE;
+    }
+
+    if (EQUALN(geomField->GetNameRef(), "OGR_GEOMETRY", strlen("OGR_GEOMETRY")))
```
why this special case ?

rouault commented on ogr/ogrsf_frmts/hana/ogrhanatablelayer.cpp:
```diff
@@ -0,0 +1,1762 @@
+        return OGRERR_FAILURE;
+    }
+
+    // TODO change an entry in attrColumns_;
```
TODO to do

mrylov replied:
The TODO is completed in 02a586c3f62ed6ef53a5d3ec00233d73acf4e3fb.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanautils.cpp:
```diff
@@ -0,0 +1,227 @@
+
+namespace hana_utils {
+
+CPLString StringFormat(const char* format, ...)
```
just use CPLString().Printf(const char* format, ...)

mrylov replied:
Done in 0a714561f809f9bb023f94eaae32471ea7acc238.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanautils.cpp:
```diff
@@ -0,0 +1,227 @@
+    default:
+        switch (wkbType)
+        {
+        case OGRwkbGeometryType::wkbPoint25D:
```
I don't think those cases can be hit as wkbFlatten(OGRwkbGeometryType::wkbPoint25D) == OGRwkbGeometryType::wkbPoint matched above

rouault commented on ogr/ogrsf_frmts/hana/ogrhanautils.cpp:
```diff
@@ -0,0 +1,227 @@
+    auto zmGeometryType = [hasZ, hasM](OGRwkbGeometryType gtype) {
+        if (hasZ)
+            return OGR_GT_SetZ(gtype);
+        if (hasM)
```
logic is flawed if both hasZ and hasM are set

rouault commented on .github/workflows/ubuntu_18.04/script.sh:
```diff
@@ -31,6 +32,9 @@ rm autotest/ogr/ogr_fgdb.py
 (cd autotest/ogr && OGR_MSSQL_CONNECTION_STRING="MSSQL:server=$IP;database=TestDB;driver=ODBC Driver 17 for SQL Server;UID=SA;PWD=DummyPassw0rd" $PYTEST ogr_mssqlspatial.py)
 
+# HANA tests
+(cd autotest/ogr && OGR_HANA_CONNECTION_STRING="DRIVER=/usr/sap/hdbclient/libodbcHDB.so;HOST=917df316-4e01-4a10-be54-eac1b6ab15fb.hana.prod-us10.hanacloud.ondemand.com;PORT=443;USER=GDALCI;PASSWORD=Z2!.S-bN]vh62\G" $PYTEST ogr_hana.py)
```
can happen in practice if I push something to a branch that is subject to a pull request: the branch is tested in my rouault/gdal fork  as a push and here in OSGeo/gdal as a PR.

mrylov replied:
Nothing should happen.
The tests are run in a separate schema which name is generated from the following query
   "SELECT REPLACE(CURRENT_UTCDATE, '-', '') || '_' || BINTOHEX(SYSUUID) FROM DUMMY;" 
The SYSUUID function returns a new universally unique identifier that is generated by the connected SAP HANA instance.   
Each time you call the SYSUUID function, it returns a new UUID value. SYSUUID calls from multiple connections are internally serialized to guarantee unique value generation.
The CURRENT_UTCDATE() call is used to track how old is the schema.

mrylov commented:
Hi Even,
Thank you very much for reviewing our pull request.
I have started working on your comments and will consider your comments step by step.

rouault commented:
> Thus, end users will get only one dynamic/shared library of the HANA driver.

There's the point I raised a few months ago regarding odbc-cpp-wrapper being used in QGIS too. There, it is embedded in QGIS itself, but if using a GDAL with SAP Hana driver enabled in QGIS, you'll have 2 instances of odbc-cpp-wrapper. While they remain the same, the duplication of symbols should be OK (perhaps restricted to the fact that there aren't global variables in odbc-cpp-wrapper), but if they start to diverge subtle bugs may arise (at least on Linux where the dynamic loader may do unintuitive things). One possibility would be that https://github.com/qgis/QGIS/tree/master/external/odbccpp/src/odbc uses a qgs::odbc namespace.

mrylov commented on .github/workflows/windows_build.yml:
```diff
@@ -264,6 +264,7 @@ jobs:
             exec { nmake /f makefile.vc check MSVC_VER=$env:MSVC_VER $env:WIN64_ARG }
             cd $env:GITHUB_WORKSPACE\autotest
             pip install -Ur requirements.txt
+            pip install -U hdbcli
```
I think the ogr_hana.py fails if the hdbcli is not installed. Hence, we should either install hdbcli or remove ogr_hana.py.

rouault replied:
ogr_hana.py already has a provision for missing hdbcli:
```
try:
    from hdbcli import dbapi
except ImportError:
    pytest.skip("hdbcli not available", allow_module_level=True)

```

mrylov replied:
Thanks, I forgot it.

The hdbcli was removed in eb4bd2e3361d6d18190f5f87c2ffa3c32cf879b9.

mrylov commented on .github/workflows/asan/start.sh:
```diff
@@ -31,11 +31,15 @@ fi
 sudo ln -s /usr/lib/ogdi/libvrf.so /usr/lib
 
+# Build odbc-cpp library for HANA
+(wget "https://cmake.org/files/v3.12/cmake-3.12.4.tar.gz" && tar xzf cmake-3.12.4.tar.gz && rm cmake-3.12.4.tar.gz && cd cmake-3.12.4 && ./bootstrap --prefix=/usr/local/ && make && make install && cd ..  && rm -rf cmake-3.12.4)
```
Done in 5ef2f19103fb45d3f3f44b3869921c2dd62961bd.

mrylov commented on .github/workflows/ubuntu_18.04/before_install.sh:
```diff
@@ -19,11 +19,39 @@ docker run --name mariadb -e MYSQL_ROOT_PASSWORD=passwd -e "MYSQL_ROOT_HOST=%" -
+  && rm hanaclient*
+export PATH=/usr/sap/hdbclient:$PATH
+
+wget "https://cmake.org/files/v3.12/cmake-3.12.4.tar.gz" \
```
Done in 5ef2f19103fb45d3f3f44b3869921c2dd62961bd and 2bc1c332a763ade25318523a214b3e167c474655.

mrylov commented on .github/workflows/ubuntu_18.04_32bit/start.sh:
```diff
@@ -65,11 +65,15 @@ sudo sh -c "cd $PWD/PROJ/build && make -j3 install"
 sudo sh -c "apt-get remove -y libproj-dev"
 
+# Build odbc-cpp library for HANA
+(wget "https://cmake.org/files/v3.12/cmake-3.12.4.tar.gz" && tar xzf cmake-3.12.4.tar.gz && rm cmake-3.12.4.tar.gz && cd cmake-3.12.4 && ./bootstrap --prefix=/usr/local/ && make && make install && cd ..  && rm -rf cmake-3.12.4)
```
Done in 5ef2f19103fb45d3f3f44b3869921c2dd62961bd.

rouault commented on .github/workflows/ubuntu_20.04/build-deps.sh:
```diff
@@ -125,4 +125,25 @@ wget -q https://github.com/rouault/pdfium_build_gdal_3_4/releases/download/v2_pd
+  && rm hanaclient*
+export PATH=/usr/sap/hdbclient:$PATH
+
+wget -q https://github.com/Kitware/CMake/releases/download/v3.12.4/cmake-3.12.4-Linux-x86_64.sh -O cmake.sh \
```
ubuntu 20.04 comes with cmake 3.16 and it is already installed by previous steps

mrylov replied:
Thank you. Removed in 4ed4eb3ab2e6510e1f98be48f0d26a4ea345c488.

mrylov commented on GDALmake.opt.in:
```diff
@@ -170,6 +170,14 @@ MYSQL_LIB  =	@MYSQL_LIB@
 MYSQL_INC  =	@MYSQL_INC@
 LIBS	   +=	$(MYSQL_LIB)
 
+#
```
That's good to know. Targeting CMake for Windows builds if fine, I'll make necessary changes according to your suggestion.

mrylov commented on gdal/configure.ac:
```diff
@@ -0,0 +1,6378 @@
+dnl ***************************************************************************
```
Does that mean that we don't need any HANA related changes in this file?

rouault replied:
I just mean that gdal/configure.ac was a file that existed in GDAL 3.4, but it has now been moved to the top directory in GDAL master 3.5.0dev. I assume your fork was started before that move.

mrylov commented on ci/travis/ubuntu_1804/script.sh:
```diff
@@ -32,6 +32,9 @@ rm autotest/ogr/ogr_fgdb.py
 (cd autotest/ogr && $PYTEST ogr_mssqlspatial.py)
 
+# HANA tests
+(cd autotest/ogr && OGR_HANA_CONNECTION_STRING="DRIVER={\'/usr/sap/hdbclient/libodbcHDB.so\'};HOST=917df316-4e01-4a10-be54-eac1b6ab15fb.hana.prod-us10.hanacloud.ondemand.com;PORT=443;USER=QGISCI;PASSWORD=tQ&7W3Klr9!p" $PYTEST ogr_hana.py)
```
@rouault, I was not able to find where the OGR_HANA_CONNECTION_STRING variable needs to be set on Ubuntu_20.04. I tried to follow OGR_MYSQL_CONNECTION_STRING/OGR_PG_CONNECTION_STRING values, but couldn't find anything. For Ubuntu_18.04, it is absolutely obvious. Could you point me to the right location?

rouault replied:
as an extra -e name=value argument to the docker run invokation of the "Run tests" step of the .yml file ?

mrylov replied:
Thank you, Even. Done in 7c3bfcaa057f85c3abfaadff43a28879c2222b0a.

mrylov commented on ogr/ogrsf_frmts/makefile.vc:
```diff
@@ -115,6 +116,15 @@ MYSQL_DIR = mysql
+PLUGIN_HANA_DIR = hana
+!ENDIF
+!ENDIF
+
```
Done in 0045a1531a4e8175a9781742c7946735840e378a.

mrylov commented on autotest/ogr/ogr_hana.py:
```diff
@@ -0,0 +1,962 @@
+###############################################################################
+# Copyright (c) 2020, SAP SE
+#
+# This library is free software; you can redistribute it and/or
```
Done in 15fc07a0dbc249fb1e4f0f6f9ebe0a0f999706ee.

mrylov commented on autotest/ogr/ogr_hana.py:
```diff
@@ -0,0 +1,962 @@
+###############################################################################
+# Initialize SAP HANA data source.
+
+def test_ogr_hana_init():
```
Done in 3e79cc6b32dc129b57b7c8e512b130c0a813cb81.

mrylov commented on autotest/ogr/ogr_hana.py:
```diff
@@ -0,0 +1,962 @@
+    conn = None
+
+    try:
+        gdal.PushErrorHandler('CPLQuietErrorHandler')
```
Done in 3e79cc6b32dc129b57b7c8e512b130c0a813cb81.

mrylov commented on autotest/ogr/ogr_hana.py:
```diff
@@ -0,0 +1,962 @@
+        gdaltest.hana_layer.CreateFeature(dst_feat)
+        feat = shp_layer.GetNextFeature()
+
+    dst_feat.Destroy()
```
Done in 3e79cc6b32dc129b57b7c8e512b130c0a813cb81.

mrylov commented on autotest/ogr/ogr_hana.py:
```diff
@@ -0,0 +1,962 @@
+
+    dst_feat.Destroy()
+
+    shp_ds.Destroy()
```
Done in 3e79cc6b32dc129b57b7c8e512b130c0a813cb81.

mrylov commented on nmake.opt:
```diff
@@ -1,1191 +1,1207 @@
-	$(INGRES_LIB) $(LIBXML2_LIB) $(PCRE_LIB) $(PCRE2_LIB) $(MONGODB_LIB_LINK) $(MONGODBV3_LIB_LINK) $(CRYPTOPP_LIB) $(OPENSSL_LIB) $(CHARLS_LIB) ws2_32.lib \
-	$(RDB_LIB) $(CRUNCH_LIB) $(OPENEXR_LIB) $(HEIF_LIB) $(LERC_LIB) $(JXL_LIB) $(BROTLI_LIB) $(BRUNSLI_LIB) $(QHULL_LIB) \
-    kernel32.lib psapi.lib wbemuuid.lib
+# $Id$
```
The changes have been reverted in f184ec815501d795c13bea73c1261308c6c35d74.

mrylov commented on configure.ac:
```diff
@@ -4044,6 +4044,66 @@ AC_SUBST(HAVE_MYSQL,$HAVE_MYSQL)
+dnl          [#include <odbc/PreparedStatement.h>], [odbc::Time time = odbc::makeNullable<odbc::time>(23, 59, 59)] -- requires static library libodbccpp.a
+           [#include<dlfcn.h>],
+           [
+             void* handle = dlopen("libodbccpp.so", RTLD_LAZY);
```
Done in b6f85d8aaac625ee259a5889204f3be1e3a3caa2.

mrylov commented on configure.ac:
```diff
@@ -4044,6 +4044,66 @@ AC_SUBST(HAVE_MYSQL,$HAVE_MYSQL)
+AC_MSG_CHECKING([for SAP HANA])
+
+if test "$ODBC_SETTING" = "no" ; then
+    AC_MSG_ERROR([SAP HANA driver requires odbc driver to be installed.])
```
Good point. Yes, the libodbccpp library requires ODBC headers.
Fixed in b6f85d8aaac625ee259a5889204f3be1e3a3caa2.

mrylov commented on doc/source/build_hints.rst:
```diff
@@ -524,6 +524,13 @@ The `GTA <https://marlam.de/gta/>`_ library is required for the :ref:`raster.gta
+HANA
+****
+
+The `odbc-cpp-wrapper library <https://github.com/SAP/odbc-cpp-wrapper>`_ is required for
```
I've enabled HANA support in cmake builds in 8f1a8ed02255806997b02691a65ca6921185f0a5.
The documentation has been adjusted accordingly.

mrylov commented on doc/source/drivers/vector/hana.rst:
```diff
@@ -0,0 +1,213 @@
+--------
+
+-  This example shows how to list HANA layers on a specified host using
+   `ogrinfo <http://www.gdal.org/ogrinfo.html>`__ command.
```
Fixed in 2f270a99c0303c1b2ba88d7092ff0feef2e00d07.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+/*                          OGRHanaDataSource                          */
+/************************************************************************/
+
+class OGRHanaDataSource : public GDALDataset
```
Done in 4c7c9b94f700fe4defbf6f0b71f54d579cb62353.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+/*                          Internal struct definitions                 */
+/************************************************************************/
+
+struct ColumnDefinition
```
Done in a5948011a6b8940389eb5f6aaca3b6ebd8aa1c67 and 44d57e10bdf1c99ce202f531475dda1d5bc5990e.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+    bool isNullable;
+    CPLString defaultValue;
+
+    AttributeColumnDescription()
```
Done in 526e3bf337ba9b9a93289d96b0fdb4cd5ddc6b43.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+    SrsCache srsCache_;
+    odbc::EnvironmentRef connEnv_;
+    odbc::ConnectionRef conn_;
+    uint majorVersion_ = 0;
```
Done in 526e3bf337ba9b9a93289d96b0fdb4cd5ddc6b43.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+    using SrsCache = std::unordered_map<int, OGRSpatialReference*>;
+
+    CPLString schemaName_;
+    bool updateMode_;
```
Done in 526e3bf337ba9b9a93289d96b0fdb4cd5ddc6b43.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+class OGRHanaLayer : public OGRLayer
+{
+protected:
+    OGRHanaDataSource* dataSource_;
```
Done in 526e3bf337ba9b9a93289d96b0fdb4cd5ddc6b43.

mrylov commented on ogr/ogrsf_frmts/hana/ogr_hana.h:
```diff
@@ -0,0 +1,367 @@
+private:
+    friend class OGRHanaDataSource;
+
+    CPLString schemaName_;
```
Done in f158a06ffc57f52806dd639c620c3a873010fd6a.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+            if (pos != tables.end())
+                tables.erase(pos);
+
+            std::unique_ptr<OGRHanaTableLayer> layer =
```
Good point. We should definitely implement this enhancement. I will consider this when the pull request is accepted.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+    return ret;
+}
+
+OGRwkbGeometryType GetGeometryType(
```
Yes, you are right. Unfortunately, there is no such metadata in HANA. What we can do is to use the LIMIT clause that of course has some drawbacks. What do you think?

rouault replied:
There's no magical solution. If for the use cases you want to address, the current solution is good enough, so be it. This should probably be documented though as a limitation in the driver doc page. And possibly add an open option where users could ask the driver to avoid this iteration and just expose wkbUnknown as the layer geometry type

mrylov replied:
Thanks for your prompt reply.
As you suggested, I've added an open options called DETECT_GEOMETRY_TYPE  that controls the layer geometry type, see a6545b9cc2d37d669cca1c1e986c46c87158237f.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+    // Note, getTableName returns correct table name also in the case
+    // when the original sql query uses a view
+    CPLString tableName = rsmd->getTableName(1);
+    if (tableName == "M_DATABASE_") // TODO: seems to be a bug
```
I've double checked this. The value of "M_DATABASE_" is set on the HANA side. Hence, we cannot do much here. Even if this is a bug that will be fixed someday, we need to leave it to support old HANA builds. I will remove the comment.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+    {
+        const char* paramOverwrite = CSLFetchNameValue(
+            options, LayerCreationOptionsConstants::OVERWRITE);
+        if (paramOverwrite != nullptr && !EQUAL(paramOverwrite, "NO"))
```
Done in a0ae1217b07388815494524824ec4f9353e62c67.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+        }
+    }
+
+    const char* paramGeomName = CSLFetchNameValue(options, LayerCreationOptionsConstants::GEOMETRY_NAME);
```
Done in a0ae1217b07388815494524824ec4f9353e62c67.

rouault commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+            srs.reset(nullptr);
+    }
+
+    srsCache_.insert({srid, srs.get()});
```
things are probably fine

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+            // return empty string if there is more than one schema.
+            if (!ret.empty())
+            {
+                ret = "";
```
Done in a0ae1217b07388815494524824ec4f9353e62c67.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+    {
+        srs = std::make_unique<OGRSpatialReference>();
+        OGRErr err = srs->importFromWkt(wkt.c_str());
+        if (OGRERR_NONE != err)
```
You're right. I've removed this in ad5f4fdde6d2f3d49deb07b2f5461a97a1166bbc.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+    try
+    {
+        CreateSpatialReferenceSystem(
+            srsLocal, authorityName, authorityCode, strWkt.c_str(),
```
The logic is changed in ad5f4fdde6d2f3d49deb07b2f5461a97a1166bbc .

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+        (srs.IsProjected()) ? srs.GetAttrValue("PROJCS")
+                            : srs.GetAttrValue("GEOGCS"));
+    if (refName.empty())
+        refName = "OGR_PROJECTION_" + std::to_string(authorityCode);
```
Reworked in ad5f4fdde6d2f3d49deb07b2f5461a97a1166bbc .

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+/*                        HasSrsPlanarEquivalent()                      */
+/************************************************************************/
+
+bool OGRHanaDataSource::HasSrsPlanarEquivalent(int srid)
```
We don't check whether the srid exists or not. Instead, we check if a planar equivalent is available in a database.

When the spatial reference system is based on a well known coordinate system, but has a different geodesic interpretation, set the srs-id value to be 1000000000 (one billion) plus the well known value. For example, the SRID for a planar interpretation of the geodetic spatial reference system WGS 84 (ID 4326) would be 1000004326.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp:
```diff
@@ -0,0 +1,1722 @@
+    }
+    rsPrimaryKeys->close();
+
+    if (std::find(ret.begin(), ret.end(), "OGR_FID") == ret.end())
```
Right, there is no guarantee. I've removed it in 9ff534523c603a3c67f797f1c88ff510350fa8d8.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanafeaturereader.cpp:
```diff
@@ -0,0 +1,512 @@
+    return (EQUAL(defaultValue, "1") || EQUAL(defaultValue, "'t'"));
+}
+
+odbc::Byte OGRHanaFeatureReader::GetFieldAsByte(int fieldIndex) const
```
Yes, it is defined as
typedef Nullable<std::int8_t>       Byte;

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanafeaturereader.cpp:
```diff
@@ -0,0 +1,512 @@
+{
+    if (IsFieldSet(fieldIndex))
+        return odbc::Long(
+            static_cast<long>(feature_.GetFieldAsInteger64(fieldIndex)));
```
odbc::Long is defined as  typedef Nullable<std::int64_t>      Long;
Hence, the cast uses a wrong type. The return value must be
return odbc::Long(feature_.GetFieldAsInteger64(fieldIndex));

mrylov replied:
Fixed in 127d206f2363e4b1f9e150341719c1a1f3301aa1.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanafeaturereader.cpp:
```diff
@@ -0,0 +1,512 @@
+{
+    if (!(std::numeric_limits<T>::min() >= value
+          || value <= std::numeric_limits<T>::max()))
+        throw "Integer value lies outside of the range";
```
It was only caught in 1 of 2 cases.
I've considered your points and adapted the implementation in 127d206f2363e4b1f9e150341719c1a1f3301aa1.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanafeaturereader.cpp:
```diff
@@ -0,0 +1,512 @@
+            else
+                throw "GetFieldAsNString: Invalid UTF-8";
+
+            int n = (seqLength <= 2) ? 1 : 2;
```
The purpose was to find the number of characters.
This weir code was simplified in 127d206f2363e4b1f9e150341719c1a1f3301aa1.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanafeaturereader.cpp:
```diff
@@ -0,0 +1,512 @@
+            &secondWithMillisecond, &timeZone);
+        double seconds = 0.0;
+        double milliseconds = modf(static_cast<double>(secondWithMillisecond), &seconds);
+        int second = static_cast<int>(nearbyint(seconds));
```
Changed in 127d206f2363e4b1f9e150341719c1a1f3301aa1.

mrylov commented on ogr/ogrsf_frmts/hana/ogrhanafeaturereader.cpp:
```diff
@@ -0,0 +1,512 @@
+        feature_.GetFieldAsDateTime(
+            fieldIndex, &year, &month, &day, &hour, &minute,
+            &secondWithMillisecond, &timeZone);
+        double seconds = 0.0;
```
Changed in 127d206f2363e4b1f9e150341719c1a1f3301aa1.

mrylov commented:
@rouault, the issue of possible symbols conflicts caused by two instances of the odbc-cpp-wrapper has been considered in https://github.com/qgis/QGIS/pull/47795. Perhaps, you want to review this pull request.

mrylov commented:
@rouault, all your comments were addressed.

rouault commented on ogr/ogrsf_frmts/generic/makefile.vc:
```diff
@@ -17,7 +17,7 @@ BASEFORMATS = -DSHAPE_ENABLED -DMITAB_ENABLED -DNTF_ENABLED -DSDTS_ENABLED -DTIG
 EXTRAFLAGS =	-I.. -I..\.. $(OGDIDEF) $(FMEDEF) $(OCIDEF) $(PGDEF) \
 		$(ODBCDEF) $(SQLITEDEF) $(MYSQLDEF) $(ILIDEF) $(DWGDEF) \
-		$(BASEFORMATS) $(IDBDEF) $(NASDEF) $(DODSDEF) \
+		$(BASEFORMATS) $(IDBDEF) $(NASDEF) $(DODSDEF) $(HANADEF) \
```
the changes in that file should be reverted

mrylov replied:
Done in 9655cc48c53c137355403d91e727ffa0539d75de.

rouault commented:
- A "HANA" line, after the MapML one, should be added in frmts/drivers.ini so that when the driver is built as a plugin, it is loaded in the same order as when it is compiled in the GDAL core library

- I gave this a local try, and I get i"Invalid read", "Conditional jump or move depends on uninitialised value" and a memory leak when ingesting a shapefile with ogr2ogr

```
$ valgrind --leak-check=full ogr2ogr -update  "HANA:$HANA_CONNECTION_STR;ENCRYPT=YES;SSL_VALIDATE_CERTIFICATE=false;CHAR_AS_UTF8=1;SCHEMA=gdal_test_20220317_9E6F323070478CCF1700AD12B99BE78A" autotest/ogr/data/poly.shp -overwrite --debug on -lco OVERWRITE=YES
==815050== Memcheck, a memory error detector
==815050== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==815050== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==815050== Command: ogr2ogr -update HANA:DRIVER=/usr/sap/hdbclient/libodbcHDB.so;HOST=917df316-4e01-4a10-be54-eac1b6ab15fb.hana.prod-us10.hanacloud.ondemand.com;PORT=443;USER=GDALCI;PASSWORD=Z2!.S-bN]vh62\\G;ENCRYPT=YES;SSL_VALIDATE_CERTIFICATE=false;CHAR_AS_UTF8=1;SCHEMA=gdal_test_20220317_9E6F323070478CCF1700AD12B99BE78A autotest/ogr/data/poly.shp -overwrite --debug on -lco OVERWRITE=YES
==815050== 
[snip]
==815050== Invalid read of size 1
==815050==    at 0x483F85C: strcasecmp (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==815050==    by 0x664224F: SetupTargetLayer::Setup(OGRLayer*, char const*, GDALVectorTranslateOptions*, long long&) (ogr2ogr_lib.cpp:4177)
==815050==    by 0x663EB6D: GDALVectorTranslate (ogr2ogr_lib.cpp:3190)
==815050==    by 0x10D344: main (ogr2ogr_bin.cpp:414)
==815050==  Address 0x1328e130 is 16 bytes inside a block of size 120 free'd
==815050==    at 0x483CFBF: operator delete(void*) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==815050==    by 0x111D57F3: __gnu_cxx::new_allocator<OGRHANA::AttributeColumnDescription>::deallocate(OGRHANA::AttributeColumnDescription*, unsigned long) (new_allocator.h:128)
==815050==    by 0x111D372B: std::allocator_traits<std::allocator<OGRHANA::AttributeColumnDescription> >::deallocate(std::allocator<OGRHANA::AttributeColumnDescription>&, OGRHANA::AttributeColumnDescription*, unsigned long) (alloc_traits.h:469)
==815050==    by 0x111D1877: std::_Vector_base<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> >::_M_deallocate(OGRHANA::AttributeColumnDescription*, unsigned long) (stl_vector.h:351)
==815050==    by 0x111D35A0: void std::vector<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> >::_M_realloc_insert<OGRHANA::AttributeColumnDescription const&>(__gnu_cxx::__normal_iterator<OGRHANA::AttributeColumnDescription*, std::vector<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> > >, OGRHANA::AttributeColumnDescription const&) (vector.tcc:500)
==815050==    by 0x111D174B: std::vector<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> >::push_back(OGRHANA::AttributeColumnDescription const&) (stl_vector.h:1195)
==815050==    by 0x111DFFF1: OGRHANA::OGRHanaTableLayer::CreateField(OGRFieldDefn*, int) (ogrhanatablelayer.cpp:1393)
==815050==    by 0x6642A29: SetupTargetLayer::Setup(OGRLayer*, char const*, GDALVectorTranslateOptions*, long long&) (ogr2ogr_lib.cpp:4273)
==815050==    by 0x663EB6D: GDALVectorTranslate (ogr2ogr_lib.cpp:3190)
==815050==    by 0x10D344: main (ogr2ogr_bin.cpp:414)
==815050==  Block was alloc'd at
==815050==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==815050==    by 0x111D7F71: __gnu_cxx::new_allocator<OGRHANA::AttributeColumnDescription>::allocate(unsigned long, void const*) (new_allocator.h:114)
==815050==    by 0x111D73BF: std::allocator_traits<std::allocator<OGRHANA::AttributeColumnDescription> >::allocate(std::allocator<OGRHANA::AttributeColumnDescription>&, unsigned long) (alloc_traits.h:443)
==815050==    by 0x111D56EB: std::_Vector_base<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> >::_M_allocate(unsigned long) (stl_vector.h:343)
==815050==    by 0x111D33F8: void std::vector<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> >::_M_realloc_insert<OGRHANA::AttributeColumnDescription const&>(__gnu_cxx::__normal_iterator<OGRHANA::AttributeColumnDescription*, std::vector<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> > >, OGRHANA::AttributeColumnDescription const&) (vector.tcc:440)
==815050==    by 0x111D174B: std::vector<OGRHANA::AttributeColumnDescription, std::allocator<OGRHANA::AttributeColumnDescription> >::push_back(OGRHANA::AttributeColumnDescription const&) (stl_vector.h:1195)
==815050==    by 0x111CE53B: OGRHANA::OGRHanaLayer::ReadFeatureDefinition(CPLString const&, CPLString const&, CPLString const&, char const*) (ogrhanalayer.cpp:723)
==815050==    by 0x111DB6EF: OGRHANA::OGRHanaTableLayer::ReadTableDefinition() (ogrhanatablelayer.cpp:421)
==815050==    by 0x111DE880: OGRHANA::OGRHanaTableLayer::Initialize(char const*, char const*) (ogrhanatablelayer.cpp:1009)
==815050==    by 0x111BAF62: OGRHanaDataSource::ICreateLayer(char const*, OGRSpatialReference*, OGRwkbGeometryType, char**) (ogrhanadatasource.cpp:1623)
==815050==    by 0x64BAEA2: GDALDataset::CreateLayer(char const*, OGRSpatialReference*, OGRwkbGeometryType, char**) (gdaldataset.cpp:4626)
==815050==    by 0x6640F19: SetupTargetLayer::Setup(OGRLayer*, char const*, GDALVectorTranslateOptions*, long long&) (ogr2ogr_lib.cpp:3848)
==815050== 
HANA: Prepare statement INSERT INTO "gdal_test_20220317_9E6F323070478CCF1700AD12B99BE78A"."POLY" ("AREA", "EAS_ID", "PRFEDEA", "OGR_GEOMETRY") VALUES(?, ?, ?, ST_GeomFromWKB(? , 27700)).
HANA: Prepare statement SELECT CURRENT_IDENTITY_VALUE() "current identity value" FROM "gdal_test_20220317_9E6F323070478CCF1700AD12B99BE78A"."POLY".
GDALVectorTranslate: 10 features written in layer 'POLY'
Shape: 10 features read on layer 'poly'.
GDAL: GDALClose(autotest/ogr/data/poly.shp, this=0x10f369e0)
==815050== Conditional jump or move depends on uninitialised value(s)
==815050==    at 0x111DDD02: OGRHANA::OGRHanaTableLayer::HasPendingFeatures() const (ogrhanatablelayer.cpp:854)
==815050==    by 0x111DDC2D: OGRHANA::OGRHanaTableLayer::FlushPendingFeatures() (ogrhanatablelayer.cpp:839)
==815050==    by 0x111DB5CD: OGRHANA::OGRHanaTableLayer::~OGRHanaTableLayer() (ogrhanatablelayer.cpp:412)
==815050==    by 0x111DB689: OGRHANA::OGRHanaTableLayer::~OGRHanaTableLayer() (ogrhanatablelayer.cpp:413)
==815050==    by 0x587786F: std::default_delete<OGRLayer>::operator()(OGRLayer*) const (unique_ptr.h:81)
==815050==    by 0x5876BC7: std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >::~unique_ptr() (unique_ptr.h:292)
==815050==    by 0x59E34FC: void std::_Destroy<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> > >(std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*) (stl_construct.h:98)
==815050==    by 0x59E2AF4: void std::_Destroy_aux<false>::__destroy<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*>(std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*, std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*) (stl_construct.h:108)
==815050==    by 0x59E204B: void std::_Destroy<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*>(std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*, std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*) (stl_construct.h:137)
==815050==    by 0x59E0FDA: void std::_Destroy<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*, std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> > >(std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*, std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*, std::allocator<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> > >&) (stl_construct.h:206)
==815050==    by 0x5D59726: std::vector<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >, std::allocator<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> > > >::_M_erase_at_end(std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >*) (stl_vector.h:1793)
==815050==    by 0x5D59665: std::vector<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> >, std::allocator<std::unique_ptr<OGRLayer, std::default_delete<OGRLayer> > > >::clear() (stl_vector.h:1496)
==815050== 
GDAL: GDALClose(HANA:DRIVER=/usr/sap/hdbclient/libodbcHDB.so;HOST=917df316-4e01-4a10-be54-eac1b6ab15fb.hana.prod-us10.hanacloud.ondemand.com;PORT=443;USER=GDALCI;PASSWORD=Z2!.S-bN]vh62\G;ENCRYPT=YES;SSL_VALIDATE_CERTIFICATE=false;CHAR_AS_UTF8=1;SCHEMA=gdal_test_20220317_9E6F323070478CCF1700AD12B99BE78A, this=0x10f443c0)
GDAL: In GDALDestroy - unloading GDAL shared library.
==815050== 
==815050== HEAP SUMMARY:
==815050==     in use at exit: 51,856 bytes in 203 blocks
==815050==   total heap usage: 80,753 allocs, 80,550 frees, 18,611,539 bytes allocated
[snip]
==815050== 867 (792 direct, 75 indirect) bytes in 9 blocks are definitely lost in loss record 162 of 176
==815050==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==815050==    by 0x111CBDFD: OGRHANA::(anonymous namespace)::CreateFieldDefn(OGRHANA::AttributeColumnDescription const&) (ogrhanalayer.cpp:183)
==815050==    by 0x111CE37E: OGRHANA::OGRHanaLayer::ReadFeatureDefinition(CPLString const&, CPLString const&, CPLString const&, char const*) (ogrhanalayer.cpp:705)
==815050==    by 0x111DB6EF: OGRHANA::OGRHanaTableLayer::ReadTableDefinition() (ogrhanatablelayer.cpp:421)
==815050==    by 0x111DE880: OGRHANA::OGRHanaTableLayer::Initialize(char const*, char const*) (ogrhanatablelayer.cpp:1009)
==815050==    by 0x111B8E63: OGRHanaDataSource::InitializeLayers(char const*, char const*)::{lambda(char const*, bool)#1}::operator()(char const*, bool) const (ogrhanadatasource.cpp:1260)
==815050==    by 0x111B915E: OGRHanaDataSource::InitializeLayers(char const*, char const*) (ogrhanadatasource.cpp:1274)
==815050==    by 0x111B572F: OGRHanaDataSource::Open(char const*, char**, int) (ogrhanadatasource.cpp:695)
==815050==    by 0x111C5F4A: OGRHanaDriverOpen(GDALOpenInfo*) (ogrhanadriver.cpp:56)
==815050==    by 0x64B90D1: GDALOpenEx (gdaldataset.cpp:3511)
==815050==    by 0x663B59B: GDALVectorTranslate (ogr2ogr_lib.cpp:2243)
==815050==    by 0x10D344: main (ogr2ogr_bin.cpp:414)
==815050== 
==815050== LEAK SUMMARY:
==815050==    definitely lost: 880 bytes in 10 blocks
```

- I also get errors when reading a layer:
```
$ valgrind --leak-check=full ogrinfo   "HANA:$HANA_CONNECTION_STR;ENCRYPT=YES;SSL_VALIDATE_CERTIFICATE=false;CHAR_AS_UTF8=1;SCHEMA=gdal_test_20220317_9E6F323070478CCF1700AD12B99BE78A" poly -al  --debug on
[...]
==815290== Conditional jump or move depends on uninitialised value(s)
==815290==    at 0x1181E89A: checkODBCNumericIndicator (GenericOutputConverter.hpp:66)
==815290==    by 0x1181E89A: lttc::enable_if<((((81u)==((Communication::Protocol::DataTypeCodeEnum)81))||((81u)==((Communication::Protocol::DataTypeCodeEnum)82)))||((81u)==((Communication::Protocol::DataTypeCodeEnum)76)))&&((18)==((SQLDBC_HostType)18)), SQLDBC_Retcode>::type SQLDBC::Conversion::convertDatabaseToHostValue<81u, 18>(SQLDBC::Conversion::DatabaseValue const&, SQLDBC::Conversion::HostValue&, SQLDBC::Conversion::ConversionOptions const&) (FixedOutputConverter.cpp:62)
==815290==    by 0x117FB925: SQLDBC::Conversion::DataConversion::convertDatabaseToHostValue(Communication::Protocol::DataTypeCodeEnum, SQLDBC::Conversion::DatabaseValue const&, SQLDBC_HostType, SQLDBC::Conversion::HostValue&, SQLDBC::Conversion::ConversionOptions const&) (DataConversion.cpp:310)
==815290==    by 0x1150077C: SQLDBC_Retcode SQLDBC::ResultSet::readAndConvertValue<SQLDBC::HeapResultSetPart>(SQLDBC::HeapResultSetPart*, unsigned int, Communication::Protocol::DataTypeCodeEnum, SQLDBC::Conversion::DatabaseValue&, long long&, SQLDBC_HostType, SQLDBC::Conversion::HostValue&, SQLDBC::Conversion::ConversionOptions const&, SQLDBC::Error&) (ResultSet.cpp:1914)
==815290==    by 0x115184CE: SQLDBC::ResultSet::readAndConvertValue(int, Communication::Protocol::DataTypeCodeEnum, SQLDBC::Conversion::DatabaseValue&, long long&, SQLDBC_HostType, SQLDBC::Conversion::HostValue&, SQLDBC::Conversion::ConversionOptions const&, SQLDBC::Error&) (ResultSet.cpp:2266)
==815290==    by 0x11522E1C: SQLDBC::RowSet::getObject(int, void*, long long, SQLDBC_HostType, long long*, bool, long long*, long, bool) (RowSet.cpp:156)
==815290==    by 0x113ABE57: SQLDBC::SQLDBC_RowSet::getObject(int, SQLDBC_HostType, void*, long long*, long long, bool) (SQLDBC.cpp:3517)
==815290==    by 0x11363E61: SQLGetData (ODBC_Statement.cpp:2715)
==815290==    by 0x73FA2C1: SQLGetData (SQLGetData.c:474)
==815290==    by 0x112925CE: odbc::ResultSet::getDecimal(unsigned short) (in /home/even/install-odbc-cpp-wrapper/lib/libodbccpp.so)
==815290==    by 0x111CD909: OGRHANA::OGRHanaLayer::ReadFeature() (ogrhanalayer.cpp:557)
==815290==    by 0x111CCD54: OGRHANA::OGRHanaLayer::GetNextFeatureInternal() (ogrhanalayer.cpp:367)
==815290==    by 0x111CF347: OGRHANA::OGRHanaLayer::GetNextFeature() (ogrhanalayer.cpp:869)
==815290== 
==815290== Invalid read of size 2
==815290==    at 0x6F03605: from_utf16_loop (loop.c:336)
==815290==    by 0x6F03605: gconv (skeleton.c:620)
==815290==    by 0x714CC7E: __gconv (gconv.c:78)
==815290==    by 0x714C56A: iconv (iconv.c:52)
==815290==    by 0x52B5033: CPLRecodeFromWCharIconv(wchar_t const*, char const*, char const*) (cpl_recode_iconv.cpp:273)
==815290==    by 0x51A202C: CPLRecodeFromWChar (cpl_recode.cpp:204)
==815290==    by 0x111CAF7D: OGRHANA::OGRHanaFeatureWriter::SetFieldValue(int, char16_t const*) (ogrhanafeaturewriter.cpp:158)
==815290==    by 0x111CDC35: OGRHANA::OGRHanaLayer::ReadFeature() (ogrhanalayer.cpp:597)
==815290==    by 0x111CCD54: OGRHANA::OGRHanaLayer::GetNextFeatureInternal() (ogrhanalayer.cpp:367)
==815290==    by 0x111CF347: OGRHANA::OGRHanaLayer::GetNextFeature() (ogrhanalayer.cpp:869)
==815290==    by 0x647F998: OGRLayer::FeatureIterator::FeatureIterator(OGRLayer*, bool) (ogrlayer.cpp:4304)
==815290==    by 0x647FBAB: OGRLayer::begin() (ogrlayer.cpp:4356)
==815290==    by 0x116C4F: begin(OGRLayer*) (ogrsf_frmts.h:287)
==815290==  Address 0x12ed28b2 is 0 bytes after a block of size 18 alloc'd
==815290==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==815290==    by 0x517D2E5: VSIMalloc (cpl_vsisimple.cpp:561)
==815290==    by 0x516241E: CPLMalloc (cpl_conv.cpp:193)
==815290==    by 0x5162393: CPLCalloc (cpl_conv.cpp:152)
==815290==    by 0x52B4E87: CPLRecodeFromWCharIconv(wchar_t const*, char const*, char const*) (cpl_recode_iconv.cpp:219)
==815290==    by 0x51A202C: CPLRecodeFromWChar (cpl_recode.cpp:204)
==815290==    by 0x111CAF7D: OGRHANA::OGRHanaFeatureWriter::SetFieldValue(int, char16_t const*) (ogrhanafeaturewriter.cpp:158)
==815290==    by 0x111CDC35: OGRHANA::OGRHanaLayer::ReadFeature() (ogrhanalayer.cpp:597)
==815290==    by 0x111CCD54: OGRHANA::OGRHanaLayer::GetNextFeatureInternal() (ogrhanalayer.cpp:367)
==815290==    by 0x111CF347: OGRHANA::OGRHanaLayer::GetNextFeature() (ogrhanalayer.cpp:869)
==815290==    by 0x647F998: OGRLayer::FeatureIterator::FeatureIterator(OGRLayer*, bool) (ogrlayer.cpp:4304)
==815290==    by 0x647FBAB: OGRLayer::begin() (ogrlayer.cpp:4356)
==815290== 
```

mrylov commented:
@rouault, I've taken all your points above into account.
- HANA is added to drivers.ini in b016de8c304a4b54160678edc9ec02abdb2bc8ee
- Fixed memory leak and invalid read in f6e6cd9ccbff2f8ad451e31253f93fcacf679637
- Fixed a conditional jump in bb651b4bd2050c7456b6d6009b43883e6a10c29e

The only issue that was not resolved is:
==815290== Conditional jump or move depends on uninitialised value(s)
==815290==    at 0x1181E89A: checkODBCNumericIndicator (GenericOutputConverter.hpp:66)
==815290==    by 0x1181E89A: lttc::enable_if<((((81u)==((Communication::Protocol::DataTypeCodeEnum)81))||((81u)==((Communication::Protocol::DataTypeCodeEnum)82)))||((81u)==((Communication::Protocol::DataTypeCodeEnum)76)))&&((18)==((SQLDBC_HostType)18)), SQLDBC_Retcode>::type SQLDBC::Conversion::convertDatabaseToHostValue<81u, 18>(SQLDBC::Conversion::DatabaseValue const&, SQLDBC::Conversion::HostValue&, SQLDBC::Conversion::ConversionOptions const&) (FixedOutputConverter.cpp:62)

To fix that I need to contact another team responsible for the HANA driver.

mrylov commented:
@rouault, I have taken all your points into account. Could you please review this pull request again?

rouault commented on doc/source/drivers/vector/hana.rst:
```diff
@@ -0,0 +1,230 @@
+-  **FID**: Specifies the name of the FID column to create. The default
+   value is 'OGR_FID'.
+-  **FID64**: Specifies whether to create the FID column with BIGINT
+   type to handle 64bit wide ids. The default value is 'OGR_FID'.
```
```suggestion
   type to handle 64bit wide ids. The default value is NO.
```

rouault commented:
ok, seems ready for me to merge.
Can you edit https://github.com/OSGeo/gdal/wiki/Maintainers-per-sub-system to add a HANA line with maintainer(s) name/github handle, or some more generic contact (not sure if everybody has edit rights, otherwise just add it here)

mrylov commented:
@rouault, thank you so much. I've added my name to the maintainers list.

rouault commented:
actually I see there's no build testing for CMake builds. Would you mind adding at least build-time testing of the driver to the build-linux-ubuntu-focal job of .github/workflows/cmake_builds.yml (and a note as a commented line that there's no runtime checking of it for now if not adding it) ? There will be some more work in a couple month to rationalize a bit all those CI setups when we'll drop nmake & autoconf support.

mrylov commented:
@rouault, I did some experiments with cmake_builds.yml, see here 11b213d0fabea0a5c3a2d3ed28744a1e3ec9afe5. Unfortunately, there is a test failure in https://github.com/mrylov/gdal/runs/5767841355?check_suite_focus=true#step:14:1092. Is it an infrastructure issue or my GA profile simply does not have required credentials?

rouault commented:
> Is it an infrastructure issue or my GA profile simply does not have required credentials?

I assume this is just a temporary unrelated failure while retrieving https://gitlab.eumetsat.int/open-source/PublicDecompWT.git for the frmts/msg driver.
Could you revert the download of CMake 3.12 ? We want to check that GDAL can build with older CMake 3.9. Given that you don't put the old CMake in the path in the step where you build odbc-cpp , you should have the CMake from the Ubuntu 20.04 distribution (3.16).

rouault commented on .github/workflows/cmake_builds.yml:
```diff
@@ -64,6 +64,23 @@ jobs:
+        git clone https://github.com/SAP/odbc-cpp-wrapper.git \
+          && mkdir odbc-cpp-wrapper/build \
+          && cd odbc-cpp-wrapper/build \
+          && cmake .. \
```
Looking at the build log:
```
2022-03-31T12:41:15.1217328Z Cloning into 'odbc-cpp-wrapper'...
2022-03-31T12:41:15.5184370Z CMake Error at CMakeLists.txt:1 (CMAKE_MINIMUM_REQUIRED):
2022-03-31T12:41:15.5185153Z   CMake 3.12 or higher is required.  You are running version 3.9.1
```

```suggestion
          && /usr/bin/cmake .. \
```

rouault commented on .github/workflows/cmake_builds.yml:
```diff
@@ -64,6 +64,23 @@ jobs:
+          && rm hanaclient*
+        export PATH=/usr/sap/hdbclient:$PATH
+        # Download and compile odbc-cpp-wrapper
+        git clone https://github.com/SAP/odbc-cpp-wrapper.git \
```
```suggestion
        sudo apt-get install -y -q cmake
        git clone https://github.com/SAP/odbc-cpp-wrapper.git \
```

mrylov commented:
@rouault, excellent! Many thanks for your help!
